R/mcmc-traces.R

```diff
@@ -304,17 +304,24 @@ mcmc_rank_overlay <- function(x,
     mutate(cut = cut(.data$value_rank, n_bins)) %>%
     group_by(.data$cut) %>%
     mutate(bin_start = min(.data$value_rank)) %>%
-    ungroup() %>%
-    select(-c("cut"))
+    ungroup()
 
+  # Count how many values fall into each bin per chain & parameter
   d_bin_counts <- data %>%
     left_join(histobins, by = "value_rank") %>%
     count(.data$parameter, .data$chain, .data$bin_start)
 
+  # Now ensure that all combinations of parameter, chain, and bin_start exist
+  # even if no counts are present (https://github.com/stan-dev/bayesplot/issues/331)
+  all_params_chains <- dplyr::distinct(data, .data$parameter, .data$chain)
+  all_bins <- dplyr::distinct(histobins, .data$bin_start, .data$cut)
+  combos <- dplyr::cross_join(all_params_chains, all_bins)
+  d_bin_counts <- full_join(combos, d_bin_counts, by = c("parameter", "chain", "bin_start")) %>%
+    mutate(n = dplyr::coalesce(n, 0L))
+
```

Comment:
There may be a more elegant solution, but this seems to get the job done

Reply:
Not sure if this is more elegant but an expand.grid with one left join and an if_else to fill the zeros should also work. Something like the following. Reads a little easier for me, though that might just be me.
  ``` 
  all_combos <- tibble::as_tibble(expand.grid(
    parameter = unique(data$parameter),
    chain = unique(data$chain),
    bin_start = unique(histobins$bin_start),
    stringsAsFactors = FALSE
  ))

  d_bin_counts <- all_combos %>%
    left_join(d_bin_counts, by = c("parameter", "chain", "bin_start")) %>%
    mutate(n = if_else(is.na(n), 0L, n))
  ```

Reply:
```
# Ensure all combinations exist and count values
d_bin_counts <- data %>%
  left_join(histobins, by = "value_rank") %>%
  count(.data$parameter, .data$chain, .data$bin_start) %>%
  right_join(
    dplyr::expand(.data$parameter, .data$chain, unique(histobins$bin_start)),
    by = c("parameter", "chain", "bin_start")
  ) %>%
  mutate(n = dplyr::coalesce(n, 0L))
  ```
  Or courtesy of claude, telling me my solution was verbouse and not dplyr-ish enough.

Reply:
Thanks to you and Claude, both!

Reply:
> dplyr::expand

Except Claude didn't realize there is no `dplyr::expand`

Reply:
Ha, great help :D